File: leetcode/src/main/kotlin/ProductExceptSelf.kt
package leetcode

class ProductExceptSelf {
    fun productExceptSelf(nums: IntArray): IntArray {

        val n = nums.size
        val res = IntArray(n) {1}
        for (i in 1 until n) {
            res[i] *= res[i-1] * nums[i-1]
        }
        var right = 1
        for (i in n-2 downTo 0) {
            right *= nums[i+1]
            res[i] *= right
        }
        return res
    }
}

File: leetcode/src/main/cpp/ProductExceptSelf.cpp
#include <bits/stdc++.h>

class Solution {
public:
    std::vector<int> productExceptSelf(std::vector<int>& nums) {
        int n = nums.size();
        std::vector<int> res(n);

        res[0] = 1;

        for (int i = 1; i < n; i++) {
            res[i] = res[i-1] * nums[i-1];
        }
        int suffix = 1;
        for (int i = n-2; i >= 0; i--) {
            suffix *= nums[i+1];
            res[i] = res[i+1] * suffix;
        }

        return res;
    }
};